Source/santametricservice/SNTMetricService.m

```diff
@@ -0,0 +1,117 @@
+/// Copyright 2021 Google Inc. All rights reserved.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///    http://www.apache.org/licenses/LICENSE-2.0
+///
+///    Unless required by applicable law or agreed to in writing, software
+///    distributed under the License is distributed on an "AS IS" BASIS,
+///    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+///    See the License for the specific language governing permissions and
+///    limitations under the License.
+
+#include <dispatch/dispatch.h>
+#include <fcntl.h>
+
+#import "Source/common/SNTConfigurator.h"
+#import "Source/common/SNTLogging.h"
+
+#import "SNTMetricService.h"
+#import "Source/santametricservice/Formats/SNTMetricRawJSONFormat.h"
+#import "Source/santametricservice/Writers/SNTMetricFileWriter.h"
+
+@interface SNTMetricService ()
+@property MOLXPCConnection *notifierConnection;
+@property MOLXPCConnection *listener;
+@property(nonatomic) dispatch_queue_t queue;
+@end
+
+@implementation SNTMetricService {
+ @private
+  SNTMetricRawJSONFormat *rawJSONFormatter;
+  NSDictionary *metricWriters;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    rawJSONFormatter = [[SNTMetricRawJSONFormat alloc] init];
+    metricWriters = @{@"file" : [[SNTMetricFileWriter alloc] init]};
+
+    _queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0);
+  }
+
+  return self;
+}
+
+/**
+ * Helper function to format NSError's for logging error messages.
+ */
+- (NSString *)messageFromError:(NSError *)error {
+  NSString *message = [error localizedDescription];
+  NSString *details = [error localizedFailureReason] ? [error localizedFailureReason] : @"";
```

Comment:
You can use the elvis operator here to avoid the duplication:

`NSString *details = [error localizedFailureReason] ?: @"";`

Reply:
Fair enough, I'd taken this from Apple's example.  So wasn't sure what was more idiomatic.  The elvis operator seems clearer.